### PR TITLE
OCM-5022: Add support for creating cluster with shared VPC

### PR DESCRIPTION
**Pull request by nirarg**
This change includes adding the following:

* "shared-vpc-policy-and-hosted-zone" module - include all vpc policy and route53 actions
* "rosa-classic-public-with-shared-vpc" example

Fixes: [OCM-5022](https://issues.redhat.com/browse/OCM-5022)

### CODE REVIEW

File: modules/account-iam-resources/versions.tf [hunk 1]
```diff
@@ -7,7 +7,7 @@ terraform {
       version = ">= 4.0"
     }
     rhcs = {
-      version = "1.4.0"
+      version = ">= 1.5.0"
```
Comment (bardielle) [inline modules/account-iam-resources/versions.tf:10]: Yay :)

File: variables.tf [hunk 2]
```diff
@@ -143,7 +143,7 @@ variable "vpc_private_subnets_ids" {
 variable "availability_zones" {
   description = "Create the vpc resources."
   type        = list(any)
-  default     = ["ap-southeast-2a", "ap-southeast-2b", "ap-southeast-2c"]
```
Comment (bardielle) [inline variables.tf:146]: :+1:

File: examples/rosa-classic-public-with-shared-vpc/main.tf [hunk 3]
```diff
@@ -0,0 +1,119 @@
+provider "aws" {
```
Comment (bardielle) [inline examples/rosa-classic-public-with-shared-vpc/main.tf:1]: Please add a README that describes all the steps for shared VPC

File: examples/rosa-classic-public-with-shared-vpc/main.tf [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+  managed = true
+}
+
+############################
+# operator roles
+############################
+module "operator_roles" {
+  source = "../../modules/operator-roles"
+  # adding a dependency to operator_policies module
```
Comment (bardielle) [inline examples/rosa-classic-public-with-shared-vpc/main.tf:64]: why? if someone will remove it what he suppose to see/get?

File: modules/shared-vpc-policy-and-hosted-zone/main.tf [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+          "route53:ChangeTagsForResource",
+          "route53:GetAccountLimit",
+          "route53:GetChange",
+          "route53:GetHostedZone",
+          "route53:ListTagsForResource",
+          "route53:UpdateHostedZoneComment",
+          "tag:GetResources",
+          "tag:UntagResources"
+        ]
+        Resource = "*"
```
Comment (bardielle) [inline modules/shared-vpc-policy-and-hosted-zone/main.tf:57]: Can you be more specific here?
I'm not sure but please check

File: examples/rosa-classic-public-with-shared-vpc/main.tf (hunk 4 above)
Comment (nirarg) [reply examples/rosa-classic-public-with-shared-vpc/main.tf:64]: Its actually not needed, will remove it

File: examples/rosa-classic-public-with-shared-vpc/main.tf [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  openshift_version            = var.openshift_version
+  replicas                     = var.replicas
+  installer_role_arn           = module.account_iam_resources.account_roles_arn["Installer"]
+  support_role_arn             = module.account_iam_resources.account_roles_arn["Support"]
+  controlplane_role_arn        = module.account_iam_resources.account_roles_arn["ControlPlane"]
+  worker_role_arn              = module.account_iam_resources.account_roles_arn["Worker"]
+  oidc_config_id               = module.oidc_provider.oidc_config_id
+  aws_subnet_ids               = module.shared-vpc-policy-and-hosted-zone.shared_subnets
+  multi_az                     = true
+  admin_credentials            = { username = "admin1", password = "123456!qwertyU" }
```
Comment (bardielle) [inline examples/rosa-classic-public-with-shared-vpc/main.tf:114]: please remove those attribute

File: modules/shared-vpc-policy-and-hosted-zone/main.tf [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  vpc {
+    vpc_id = var.vpc_id
+  }
+  lifecycle {
+    ignore_changes = [tags]
+  }
+}
+
+# This resource is used in order to avoid any of the resources triggers here
+# from being destroyed before the cluster distruction is done
```
Comment (bardielle) [inline modules/shared-vpc-policy-and-hosted-zone/main.tf:101]: Please add one more sentence about the "regular dependency"

File: modules/rosa-cluster-classic/main.tf [hunk 8]
```diff
@@ -26,7 +32,7 @@ resource "rhcs_cluster_rosa_classic" "rosa_sts_cluster" {
   }
   sts                = local.sts_roles
   aws_subnet_ids     = var.aws_subnet_ids
-  availability_zones = var.availability_zones
+  availability_zones = distinct(data.aws_subnet.provided_subnet[*].availability_zone)
   // aws_private_link   = var.aws_private_link
```
Comment (bardielle) [inline modules/rosa-cluster-classic/main.tf:36]: remove those comments 
and keep the attribute of availability_zones

File: examples/rosa-classic-public-with-shared-vpc/main.tf (hunk 3 above)
Comment (nirarg) [reply examples/rosa-classic-public-with-shared-vpc/main.tf:1]: Will do it in other PR, as par of the work on [OCM-4992](https://issues.redhat.com/browse/OCM-4992)

File: examples/rosa-classic-public-with-shared-vpc/main.tf (hunk 4 above)
Comment (nirarg) [reply examples/rosa-classic-public-with-shared-vpc/main.tf]: I removed this dependency and added time_wait resources to the relevant modules

File: modules/shared-vpc-policy-and-hosted-zone/main.tf (hunk 5 above)
Comment (nirarg) [reply modules/shared-vpc-policy-and-hosted-zone/main.tf:63]: I change it a bit
Anyway, we will have to review this to understand if this is the minimal permissions can be added
Opened ticket [OCM-5663](https://issues.redhat.com/browse/OCM-5663)

File: examples/rosa-classic-public-with-shared-vpc/main.tf (hunk 6 above)
Comment (nirarg) [reply examples/rosa-classic-public-with-shared-vpc/main.tf]: Changed it to use `random_password` instead

File: modules/rosa-cluster-classic/main.tf (hunk 8 above)
Comment (nirarg) [reply modules/rosa-cluster-classic/main.tf]: Done

**Comment by bardielle** [comment]
Great work!!

/lgtm
/approve

**Pull request merged**